pkg/flavor/flavor.go

```diff
@@ -30,27 +34,105 @@ func (c Conflicts) String() string {
 }
 
 type methods struct {
-	flavorStore        cache.Store
-	clusterFlavorStore cache.Store
+	flavorStore            cache.Store
+	clusterFlavorStore     cache.Store
+	preferenceStore        cache.Store
+	clusterPreferenceStore cache.Store
 }
 
 var _ Methods = &methods{}
 
-func NewMethods(flavorStore, clusterFlavorStore cache.Store) Methods {
+func NewMethods(flavorStore, clusterFlavorStore, preferenceStore, clusterPreferenceStore cache.Store) Methods {
 	return &methods{
-		flavorStore:        flavorStore,
-		clusterFlavorStore: clusterFlavorStore,
+		flavorStore:            flavorStore,
+		clusterFlavorStore:     clusterFlavorStore,
+		preferenceStore:        preferenceStore,
+		clusterPreferenceStore: clusterPreferenceStore,
 	}
 }
 
-func (m *methods) ApplyToVmi(field *k8sfield.Path, flavorSpec *flavorv1alpha1.VirtualMachineFlavorSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) Conflicts {
-
+func (m *methods) ApplyToVmi(field *k8sfield.Path, flavorSpec *flavorv1alpha1.VirtualMachineFlavorSpec, preferenceSpec *flavorv1alpha1.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) Conflicts {
 	var conflicts Conflicts
 
-	conflicts = append(conflicts, applyCpu(field, flavorSpec, vmiSpec)...)
+	if flavorSpec != nil {
+		conflicts = append(conflicts, applyCpu(field, flavorSpec, preferenceSpec, vmiSpec)...)
+		conflicts = append(conflicts, applyMemory(field, flavorSpec, vmiSpec)...)
+		conflicts = append(conflicts, applyIOThreadPolicy(field, flavorSpec, vmiSpec)...)
+		conflicts = append(conflicts, applyLaunchSecurity(field, flavorSpec, vmiSpec)...)
+		conflicts = append(conflicts, applyGPUs(field, flavorSpec, vmiSpec)...)
+		conflicts = append(conflicts, applyHostDevices(field, flavorSpec, vmiSpec)...)
+	}
+
+	if preferenceSpec != nil {
+		// By design Preferences can't conflict with the VMI so we don't return any
+		applyDevicePreferences(preferenceSpec, vmiSpec)
+		applyFeaturePreferences(preferenceSpec, vmiSpec)
+		applyFirmwarePreferences(preferenceSpec, vmiSpec)
+		applyMachinePreferences(preferenceSpec, vmiSpec)
+		applyClockPreferences(preferenceSpec, vmiSpec)
+	}
 
 	return conflicts
+}
+
+func (m *methods) FindPreferenceSpec(vm *virtv1.VirtualMachine) (*flavorv1alpha1.VirtualMachinePreferenceSpec, error) {
+
+	if vm.Spec.Preference == nil {
+		return nil, nil
+	}
+
+	switch strings.ToLower(vm.Spec.Preference.Kind) {
+	case apiflavor.SingularPreferenceResourceName, apiflavor.PluralPreferenceResourceName:
+		preference, err := m.findPreference(vm)
+		if err != nil {
+			return nil, err
+		}
+		return &preference.Spec, nil
+	case apiflavor.ClusterSingularPreferenceResourceName, apiflavor.ClusterPluralPreferenceResourceName, "":
```

Comment:
IMO it's better to change the empty kind string (`""`) to a default kind in mutating webhooks. This way you don't have to make assumptions that an empty string means something else. It's also much easier to change the defaults in the future that way.

Reply:
Coming back to this, I forgot that we are actually calling this initially in the validating webhook, that's called before the mutating webhook right?

https://github.com/kubevirt/kubevirt/blob/a1f14b2901ea34c5f090b8233580bd42989cc49f/pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go#L169-L186

Reply:
Apologies, I had this the wrong way around:

https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#what-are-admission-webhooks

Reply:
Okay hopefully fixed by https://github.com/kubevirt/kubevirt/pull/7578/commits/8ab9f209dbca6a07b50f41bb393d8e2228a77ea6

Reply:
Awesome :)